README.md
# `use-package`

[![Join the chat at https://gitter.im/use-package/Lobby](https://badges.gitter.im/use-package/Lobby.svg)](https://gitter.im/use-package/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status](https://github.com/jwiegley/use-package/actions/workflows/test.yml/badge.svg)](https://github.com/jwiegley/use-package/actions)
[![GNU ELPA](https://elpa.gnu.org/packages/use-package.svg)](https://elpa.gnu.org/packages/use-package.html)

The `use-package` macro allows you to isolate package configuration in your
`.emacs` file in a way that is both performance-oriented and, well, tidy.  I
created it because I have over 80 packages that I use in Emacs, and things
were getting difficult to manage.  Yet with this utility my total load time is
around 2 seconds, with no loss of functionality!

**NOTE**: `use-package` is **not** a package manager! Although `use-package`
does have the useful capability to interface with package managers (see
[below](#package-installation)), its primary purpose is for the configuration
and loading of packages.

- [Installing use-package](#installing-use-package)
- [Getting started](#getting-started)
- [Key-binding](#key-binding)
	+ [Binding to keymaps](#binding-to-keymaps)
	+ [Binding within local keymaps](#binding-within-local-keymaps)
- [Modes and interpreters](#modes-and-interpreters)
- [Magic handlers](#magic-handlers)
- [Hooks](#hooks)
- [Package customization](#package-customization)
  + [Customizing variables](#customizing-variables)
  + [Customizing faces](#customizing-faces)
- [Notes about lazy loading](#notes-about-lazy-loading)
- [Information about package loads](#information-about-package-loads)
- [Conditional loading](#conditional-loading)
	+ [Conditional loading before :preface](#conditional-loading-before-preface)
	+ [Loading packages in a sequence](#loading-packages-in-sequence)
	+ [Prevent loading if dependencies are missing](#prevent-loading-if-dependencies-are-missing)
- [Byte compiling your .emacs](#byte-compiling-your-emacs)
	+ [Prevent a package from loading at compile-time](#prevent-a-package-from-loading-at-compile-time)
- [Extending the load-path](#extending-the-load-path)
- [Catching errors during use-package expansion](#catching-errors-during-use-package-expansion)
- [Diminishing and delighting minor modes](#diminishing-and-delighting-minor-modes)
- [Package installation](#package-installation)
	+ [Usage with other package managers](#usage-with-other-package-managers)
- [Gathering Statistics](#gathering-statistics)
- [Keyword Extensions](#keyword-extensions)
	+ [use-package-ensure-system-package](#use-package-ensure-system-package)
	+ [use-package-chords](#use-package-chords)
	+ [How to create an extension](#how-to-create-an-extension)
		+ [First step: Add the keyword](#first-step-add-the-keyword)
		+ [Second step: Create a normalizer](#second-step-create-a-normalizer)
		+ [Third step: Create a handler](#third-step-create-a-handler)
		+ [Fourth step: Test it out](#fourth-step-test-it-out)
- [Some timing results](#some-timing-results)
* [Upgrading to 2.x](#upgrading-to-2x)
	+ [Semantics of :init is now consistent](#semantics-of-init-is-now-consistent)
	+ [:idle has been removed](#idle-has-been-removed)
	+ [:defer now accepts an optional numeric argument](#defer-now-accepts-an-optional-numeric-argument)
	+ [Add :preface, occurring before everything except :disabled](#add-preface-occurring-before-everything-except-disabled)
	+ [Add :functions, for declaring functions to the byte-compiler](#add-functions-for-declaring-functions-to-the-byte-compiler)
	+ [use-package.el is no longer needed at runtime](#use-packageel-is-no-longer-needed-at-runtime)
## Installing use-package

Either clone from this GitHub repository or install from
[GNU ELPA](https://elpa.gnu.org/) (recommended).

## Getting started

Here is the simplest `use-package` declaration:

``` elisp
;; This is only needed once, near the top of the file
(eval-when-compile
  ;; Following line is not needed if use-package.el is in ~/.emacs.d
  (add-to-list 'load-path "<path where use-package is installed>")
  (require 'use-package))

(use-package foo)
```

This loads in the package `foo`, but only if `foo` is available on your
system. If not, a warning is logged to the `*Messages*` buffer.

Use the `:init` keyword to execute code before a package is loaded.  It
accepts one or more forms, up to the next keyword:

``` elisp
(use-package foo
  :init
  (setq foo-variable t))
```

Similarly, `:config` can be used to execute code after a package is loaded.
In cases where loading is done lazily (see more about autoloading below), this
execution is deferred until after the autoload occurs:

``` elisp
(use-package foo
  :init
  (setq foo-variable t)
  :config
  (foo-mode 1))
```

As you might expect, you can use `:init` and `:config` together:

``` elisp
(use-package color-moccur
  :commands (isearch-moccur isearch-all)
  :bind (("M-s O" . moccur)
         :map isearch-mode-map
         ("M-o" . isearch-moccur)
         ("M-O" . isearch-moccur-all))
  :init
  (setq isearch-lazy-highlight t)
  :config
  (use-package moccur-edit))
```

In this case, I want to autoload the commands `isearch-moccur` and
`isearch-all` from `color-moccur.el`, and bind keys both at the global level
and within the `isearch-mode-map` (see next section).  When the package is
actually loaded (by using one of these commands), `moccur-edit` is also
loaded, to allow editing of the `moccur` buffer.

If you autoload non-interactive function, please use `:autoload`.

```elisp
(use-package org-crypt
  :autoload org-crypt-use-before-save-magic)
```

## Key-binding

Another common thing to do when loading a module is to bind a key to primary
commands within that module:

``` elisp
(use-package ace-jump-mode
  :bind ("C-." . ace-jump-mode))
```

This does two things: first, it creates an autoload for the `ace-jump-mode`
command and defers loading of `ace-jump-mode` until you actually use it.
Second, it binds the key `C-.` to that command.  After loading, you can use
`M-x describe-personal-keybindings` to see all such keybindings you've set
throughout your `.emacs` file.

A more literal way to do the exact same thing is:

``` elisp
(use-package ace-jump-mode
  :commands ace-jump-mode
  :init
  (bind-key "C-." 'ace-jump-mode))
```

When you use the `:commands` keyword, it creates autoloads for those commands
and defers loading of the module until they are used.  Since the `:init` form
is always run -- even if `ace-jump-mode` might not be on your system --
remember to restrict `:init` code to only what would succeed either way.

The `:bind` keyword takes either a cons or a list of conses:

``` elisp
(use-package hi-lock
  :bind (("M-o l" . highlight-lines-matching-regexp)
         ("M-o r" . highlight-regexp)
         ("M-o w" . highlight-phrase)))
```

Alternatively, the command name may be replaced with a cons `(desc . command)`,
where `desc` is a string describing `command`, which is the name of a command
to bind to:

```elisp
(use-package avy
  :bind ("C-:" ("Jump to char" . avy-goto-char)
         "M-g f" ("Jump to line" . avy-goto-line)))
```

These descriptions can be used by other code that deals with key bindings.
For example, the GNU ELPA package which-key displays them when showing key
bindings, instead of the plain command names.

The `:commands` keyword takes either a symbol or a list of symbols.

**NOTE**: inside strings, special keys like `tab` or `F1`-`Fn` have to be written inside angle brackets, e.g. `"C-<up>"`.
Standalone special keys (and some combinations) can be written in square brackets, e.g. `[tab]` instead of `"<tab>"`. The syntax for the keybindings is similar to
the "kbd" syntax: see [https://www.gnu.org/software/emacs/manual/html_node/emacs/Init-Rebinding.html](https://www.gnu.org/software/emacs/manual/html_node/emacs/Init-Rebinding.html)
for more information.

Examples:

``` elisp
(use-package helm
  :bind (("M-x" . helm-M-x)
         ("M-<f5>" . helm-find-files)
         ([f10] . helm-buffers-list)
         ([S-f10] . helm-recentf)))
```

Furthermore, [remapping commands](https://www.gnu.org/software/emacs/manual/html_node/elisp/Remapping-Commands.html)
with `:bind` and `bind-key` works as expected, because when the
binding is a vector, it is passed straight to `define-key`. So the
following example will rebind `M-q` (originally `fill-paragraph`) to
`unfill-toggle`:

``` elisp
(use-package unfill
  :bind ([remap fill-paragraph] . unfill-toggle))
```

### Binding to keymaps

Normally `:bind` expects that commands are functions that will be autoloaded
from the given package. However, this does not work if one of those commands
is actually a keymap, since keymaps are not functions, and cannot be
autoloaded using Emacs' `autoload` mechanism.

To handle this case, `use-package` offers a special, limited variant of
`:bind` called `:bind-keymap`. The only difference is that the "commands"
bound to by `:bind-keymap` must be keymaps defined in the package, rather than
command functions. This is handled behind the scenes by generating custom code
that loads the package containing the keymap, and then re-executes your
keypress after the first load, to reinterpret that keypress as a prefix key.

For example:

``` elisp
(use-package projectile
  :bind-keymap
  ("C-c p" . projectile-command-map))
```

### Binding within local keymaps

Slightly different from binding a key to a keymap, is binding a key *within* a
local keymap that only exists after the package is loaded.  `use-package`
supports this with a `:map` modifier, taking the local keymap to bind to:

``` elisp
(use-package helm
  :bind (:map helm-command-map
         ("C-c h" . helm-execute-persistent-action)))
```

The effect of this statement is to wait until `helm` has loaded, and then to
bind the key `C-c h` to `helm-execute-persistent-action` within Helm's local
keymap, `helm-command-map`.

Multiple uses of `:map` may be specified. Any binding occurring before the
first use of `:map` are applied to the global keymap:

``` elisp
(use-package term
  :bind (("C-c t" . term)
         :map term-mode-map
         ("M-p" . term-send-up)
         ("M-n" . term-send-down)
         :map term-raw-map
         ("M-o" . other-window)
         ("M-p" . term-send-up)
         ("M-n" . term-send-down)))
```
### Binding to repeat-maps

A special case of binding within a local keymap is when that keymap is
used by `repeat-mode`. These keymaps are usually defined specifically
for this. Using the `:repeat-map` keyword, and passing it a name for
the map it defines, will bind all following keys inside that map, and
(by default) set the `repeat-map` property of each bound command to
that map.


This creates a keymap called `git-gutter+-repeat-map`, makes four
bindings in it as above, then sets the `repeat-map` property of each
bound command (`git-gutter+-next-hunk` `git-gutter+-previous-hunk`,
`git-gutter+-stage-hunks` and `git-gutter+-revert-hunk`) to that
keymap.

``` elisp
(use-package git-gutter+
  :bind
  (:repeat-map git-gutter+-repeat-map
   ("n" . git-gutter+-next-hunk)
   ("p" . git-gutter+-previous-hunk)
   ("s" . git-gutter+-stage-hunks)
   ("r" . git-gutter+-revert-hunk)))
```

Specifying `:exit` inside the scope of `:repeat-map` will prevent the
`repeat-map` property being set, so that the command can be used from
within the repeat map, but after it using it the repeat map will no
longer be available. This is useful for commands often used at the end
of a series of repeated commands:

``` elisp
(use-package git-gutter+
  :bind
  (:repeat-map my/git-gutter+-repeat-map
   ("n" . git-gutter+-next-hunk)
   ("p" . git-gutter+-previous-hunk)
   ("s" . git-gutter+-stage-hunks)
   ("r" . git-gutter+-revert-hunk)
   :exit
   ("c" . magit-commit-create)
   ("C" . magit-commit)
   ("b" . magit-blame)))
```

Specifying `:continue` *forces* setting the `repeat-map` property
(just like *not* specifying `:exit`), so these two snippets are
equivalent:

``` elisp
(use-package git-gutter+
  :bind
  (:repeat-map my/git-gutter+-repeat-map
   ("n" . git-gutter+-next-hunk)
   ("p" . git-gutter+-previous-hunk)
   ("s" . git-gutter+-stage-hunks)
   ("r" . git-gutter+-revert-hunk)
   :exit
   ("c" . magit-commit-create)
   ("C" . magit-commit)
   ("b" . magit-blame)))
```

``` elisp
(use-package git-gutter+
  :bind
  (:repeat-map my/git-gutter+-repeat-map
   :exit
   ("c" . magit-commit-create)
   ("C" . magit-commit)
   ("b" . magit-blame)
   :continue
   ("n" . git-gutter+-next-hunk)
   ("p" . git-gutter+-previous-hunk)
   ("s" . git-gutter+-stage-hunks)
   ("r" . git-gutter+-revert-hunk)))
```
   



## Modes and interpreters

Similar to `:bind`, you can use `:mode` and `:interpreter` to establish a
deferred binding within the `auto-mode-alist` and `interpreter-mode-alist`
variables. The specifier to either keyword can be a cons cell, a list of cons
cells, or a string or regexp:

``` elisp
(use-package ruby-mode
  :mode "\\.rb\\'"
  :interpreter "ruby")

;; The package is "python" but the mode is "python-mode":
(use-package python
  :mode ("\\.py\\'" . python-mode)
  :interpreter ("python" . python-mode))
```

If you aren't using `:commands`, `:bind`, `:bind*`, `:bind-keymap`,
`:bind-keymap*`, `:mode`, `:interpreter`, or `:hook` (all of which imply `:defer`; see
the docstring for `use-package` for a brief description of each), you can
still defer loading with the `:defer` keyword:

``` elisp
(use-package ace-jump-mode
  :defer t
  :init
  (autoload 'ace-jump-mode "ace-jump-mode" nil t)
  (bind-key "C-." 'ace-jump-mode))
```

This does exactly the same thing as the following:

``` elisp
(use-package ace-jump-mode
  :bind ("C-." . ace-jump-mode))
```

## Magic handlers

Similar to `:mode` and `:interpreter`, you can also use `:magic` and
`:magic-fallback` to cause certain function to be run if the beginning of a
file matches a given regular expression. The difference between the two is
that `:magic-fallback` has a lower priority than `:mode`. For example:

``` elisp
(use-package pdf-tools
  :load-path "site-lisp/pdf-tools/lisp"
  :magic ("%PDF" . pdf-view-mode)
  :config
  (pdf-tools-install :no-query))
```

This registers an autoloaded command for `pdf-view-mode`, defers loading of
`pdf-tools`, and runs `pdf-view-mode` if the beginning of a buffer matches the
string `"%PDF"`.

## Hooks

The `:hook` keyword allows adding functions onto package hooks. The
following are equivalent:

``` elisp
(use-package company
  :hook (prog-mode . company-mode))

(use-package company
  :commands company-mode
  :init
  (add-hook 'prog-mode-hook #'company-mode))
```

And likewise, when multiple hooks should be applied, all of the
following are also equivalent:

``` elisp
(use-package company
  :hook ((prog-mode text-mode) . company-mode))

(use-package company
  :hook ((prog-mode . company-mode)
         (text-mode . company-mode)))

(use-package company
  :hook (prog-mode . company-mode)
  :hook (text-mode . company-mode))

(use-package company
  :hook
  (prog-mode . company-mode)
  (text-mode . company-mode))

(use-package company
  :commands company-mode
  :init
  (add-hook 'prog-mode-hook #'company-mode)
  (add-hook 'text-mode-hook #'company-mode))
```

When using `:hook`, omit the "-hook" suffix if you specify the hook
explicitly, as this is appended by default. For example, the following
code will not work as it attempts to add to the `prog-mode-hook-hook`
which does not exist:

``` elisp
;; DOES NOT WORK
(use-package ace-jump-mode
  :hook (prog-mode-hook . ace-jump-mode))
```

If you do not like this behaviour, set `use-package-hook-name-suffix`
to nil. By default the value of this variable is "-hook".

The use of `:hook`, as with `:bind`, `:mode`, `:interpreter`, etc., causes the
functions being hooked to implicitly be read as `:commands` (meaning they will
establish interactive `autoload` definitions for that module, if not already
defined as functions), and so `:defer t` is also implied by `:hook`.

## Package customization

### Customizing variables.

The `:custom` keyword allows customization of package custom variables.

``` elisp
(use-package comint
  :custom
  (comint-buffer-maximum-size 20000 "Increase comint buffer size.")
  (comint-prompt-read-only t "Make the prompt read only."))
```

The documentation string is not mandatory.

**NOTE**: these are only for people who wish to keep customizations with their
accompanying use-package declarations. Functionally, the only benefit over
using `setq` in a `:config` block is that customizations might execute code
when values are assigned.

**NOTE**: The customized values are **not** saved in the Emacs `custom-file`.
Thus you should either use the `:custom` option **or** you should use `M-x
customize-option` which will save customized values in the Emacs `custom-file`.
Do not use both.

### Customizing faces

The `:custom-face` keyword allows customization of package custom faces.

``` elisp
(use-package eruby-mode
  :custom-face
  (eruby-standard-face ((t (:slant italic)))))

(use-package example
  :custom-face
  (example-1-face ((t (:foreground "LightPink"))))
  (example-2-face ((t (:foreground "LightGreen"))) face-defface-spec))

(use-package zenburn-theme
  :preface
  (setq my/zenburn-colors-alist
        '((fg . "#DCDCCC") (bg . "#1C1C1C") (cyan . "#93E0E3")))
  :custom-face
  (region ((t (:background ,(alist-get my/zenburn-colors-alist 'cyan)))))
  :config
  (load-theme 'zenburn t))
```

## Notes about lazy loading

The keywords `:commands`, et al, provide "triggers" that cause a package to be
loaded when certain events occur. However, if `use-package` cannot determine
that any trigger has been declared, it will load the package immediately (when
Emacs is starting up) unless `:defer t` is given. The presence of triggers can
be overridden using `:demand t` to force immediately loading anyway. For
example, `:hook` represents a trigger that fires when the specified hook is
run.

In almost all cases you don't need to manually specify `:defer t`, because
this is implied whenever `:bind` or `:mode` or `:interpreter` are used.
Typically, you only need to specify `:defer` if you know for a fact that some
other package will do something to cause your package to load at the
appropriate time, and thus you would like to defer loading even though
`use-package` has not created any autoloads for you.

You can override package deferral with the `:demand` keyword. Thus, even if
you use `:bind`, adding `:demand` will force loading to occur immediately and
not establish an autoload for the bound key.

## Information about package loads

When a package is loaded, and if you have `use-package-verbose` set to `t`, or
if the package takes longer than 0.1s to load, you will see a message to
indicate this loading activity in the `*Messages*` buffer.  The same will
happen for configuration, or `:config` blocks that take longer than 0.1s to
execute.  In general, you should keep `:init` forms as simple and quick as
possible, and put as much as you can get away with into the `:config` block.
This way, deferred loading can help your Emacs to start as quickly as
possible.

Additionally, if an error occurs while initializing or configuring a package,
this will not stop your Emacs from loading.  Rather, the error will be
captured by `use-package`, and reported to a special `*Warnings*` popup
buffer, so that you can debug the situation in an otherwise functional Emacs.

## Conditional loading

You can use the `:if` keyword to predicate the loading and initialization of
modules.

For example, I only want `edit-server` running for my main,
graphical Emacs, not for other Emacsen I may start at the command line:

``` elisp
(use-package edit-server
  :if window-system
  :init
  (add-hook 'after-init-hook 'server-start t)
  (add-hook 'after-init-hook 'edit-server-start t))
```
In another example, we can load things conditional on the operating system:

``` elisp
(use-package exec-path-from-shell
  :if (memq window-system '(mac ns))
  :ensure t
  :config
  (exec-path-from-shell-initialize))
```

The `:disabled` keyword can turn off a module you're having difficulties with,
or stop loading something you're not using at the present time:

``` elisp
(use-package ess-site
  :disabled
  :commands R)
```

When byte-compiling your `.emacs` file, disabled declarations are omitted
from the output entirely, to accelerate startup times.

**NOTE**: `:when` is provided as an alias for `:if`, and `:unless foo` means
the same thing as `:if (not foo)`.

### Conditional loading before :preface

If you need to conditionalize a use-package form so that the condition occurs
before even the `:preface` is executed, simply use `when` around the
use-package form itself.  For example, the following will also stop
`:ensure` from happening on Mac systems:

``` elisp
(when (memq window-system '(mac ns))
  (use-package exec-path-from-shell
    :ensure t
    :config
    (exec-path-from-shell-initialize)))
```

### Loading packages in sequence

Sometimes it only makes sense to configure a package after another has been
loaded, because certain variables or functions are not in scope until that
time. This can achieved using an `:after` keyword that allows a fairly rich
description of the exact conditions when loading should occur. Here is an
example:

``` elisp
(use-package hydra
  :load-path "site-lisp/hydra")

(use-package ivy
  :load-path "site-lisp/swiper")

(use-package ivy-hydra
  :after (ivy hydra))
```

In this case, because all of these packages are demand-loaded in the order
they occur, the use of `:after` is not strictly necessary. By using it,

NEWS.md
# Changes

## 2.4.5

This is mostly a bug-fix release:

- When `:hook` is using in the form `(use-package foo :hook bar)`,
  this is once again equivalent to
  `(use-package foo :hook (bar . foo))` and no longer the same as
  `(use-package foo :hook (bar . foo-mode))`.

## 2.4.4

This release prepares for inclusion to GNU ELPA and includes no other changes

## 2.4.1

This is mostly a bug-fix release:

- Update the documentation for :custom as per #850

- Fix broken test due to #850

- better tests

- add test for #845

- Support keymap symbol in bind-key. Fix #845

- use-package-core.el: use the Emacs set-default function to avoid saving :custom vars twice

- Fix Travis

- typo, should be a vector, not a bytecode object

  Solves https://github.com/jwiegley/use-package/issues/842

- Add special value back again, in case needed for backwards compat

  I don't know why this special value exists, but perhaps old client code uses it.

  The additional `t' in the macro expansion is accidental but not harmful I guess.

- Even when there's no :config, run any pre/post config hooks

  i.e., following the existing docs for use-package-inject-hooks, these hooks are
  run:

  use-package--foo--pre-config-hook
  use-package--foo--post-config-hook

  This should make config customisations more predictable (for example, spacemacs
  uses these hooks extensively to allow 'layers' to be customised).

  I got rid of the "special" default value for :config, because it doesn't seem to
  be treated any differently than nil.

  Fixes #785

- Clarify the documentation for :after

- add table of contents to README

- Fix typos

  Typos found with codespell.

- Fix typos

- Attempt to explain omit "-hook" better

- Update tests

- Switch from `require' to `load' + `featurep'

- Use `require', not `load', when byte-compiling

- Make custom-face evaluate elisp.

  Fix #696.

- Add a line of documentation for (use-pacakage ... :hook).

- Fix typo in README

- Fix documentation for defer

- Add no-query option for pdf-tools-install

- Fix typo in README

- Fix all notes in README

- Mention use-package-ensure in README

  Without requiring `use-package-ensure`, setting `use-package-always-ensure`
  did not actually work for me.

## 2.4

### Breaking changes

- `use-package` no longer requires `diminish` as a dependency, allowing people
  to decide whether they want to use diminish or delight. This means that if
  you do use diminish, you'll now need to pull it into your configuration
  before any use of the `:diminish` kewyord. For example:

  ``` elisp
      (use-package diminish :ensure t)
  ```

- Emacs 24.3 or higher is now a requirement.

- The `:defer-install` keyword has been removed. It may reappear as an add-on
  module for use-package in a future release. See issue #442 for more details.

- There is no longer a `use-package-debug` option, since `use-package-verbose`
  already has the possible value of `debug`.

- The ordering of several elements of `use-package-keywords` have changed; if
  you had previously customized this (or were an extension author adding to
  this list), you may need to rework your changes.

- For extension authors, `:commands` should no longer be propagated down for
  autoloading. See more below.

### Other changes

- Upgrade license to GPL 3.

- If `use-package-verbose` is set to the symbol `debug`, any evaluation errors
  during package configuration will cause a complete report to be written to a
  `*use-package*` buffer, including: the text of the error, the `use-package`
  declaration that caused the error, the post-normalized form of this
  declaration, and the macro-expanded version (without verbosity-related
  code). Note that this still does not help if there are parsing errors, which
  cause Emacs to register a Lisp error at startup time.

- New customization variable `use-package-deferring-keywords`, mainly intended
  for use by extension packages, indicates keywords that, if used without
  `:demand`, cause deferred loading (as if `:defer t` had been specified).

- The `:ensure` keyword now accepts a specific pinning sub-keyword. For
  example:

  ``` elisp
  (use-package foo
    :pin "elpa")
  ```

  This ensure the package `foo` is installed from `"elpa"`.

  ``` elisp
  (use-package foo
    :ensure bar
    :ensure (quux :pin "melpa"))
  ```

  This says that `foo` ensures that `bar` is installed, as well as `quux` from
  `"melpa"`. It does *not* ensure that `foo` is installed, because explicit
  `:ensure` keywords were given.

- New `:hook` keyword.

- New `:catch` keyword. If `t` or `nil`, it enables (the default, see
  `use-package-defaults`) or disables catching errors at load time in
  use-package expansions. It can also be a function taking two arguments: the
  keyword being processed at the time the error was encountered, and the error
  object (as generated by `condition-case`).

- New keywords `:custom (foo1 bar1) (foo2 bar2)` etc., and `:custom-face`.

  NOTE: These are only for people who wish to keep customizations with their
  accompanying use-package declarations. Functionally, the only benefit over
  using `setq` in a `:config` block is that customizations might execute code
  when values are assigned. If you currently use `M-x customize-option` and
  save to a settings file, you do not want to use this option.

- New `:magic` and `:magic-fallback` keywords.

- New `:defer-install` keyword.

- New customization variable `use-package-enable-imenu-support`.

- New customization variable `use-package-hook-name-suffix`. Any symbols named
  in `:hook`, or in the CAR of cons cells passed to `:hook`, have this text
  appended to them as a convenience. If you find yourself using this keyword
  to add to hooks of different names, or just don't want such appending done,
  you can change the text to an empty string.

- New customization variable `use-package-compute-statistics`, and an
  accompanying command `M-x use-package-report`. See the README for more

use-package-ensure.el
;;; use-package-ensure.el --- Support for the :ensure and :pin keywords

;; Copyright (C) 2012-2017 John Wiegley

;; Author: John Wiegley <johnw@newartisans.com>
;; Maintainer: John Wiegley <johnw@newartisans.com>
;; Created: 17 Jun 2012
;; Modified: 3 Dec 2017
;; Version: 1.0
;; Package-Requires: ((emacs "24.3") (use-package "2.4"))
;; Keywords: dotemacs startup speed config package
;; URL: https://github.com/jwiegley/use-package

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License as
;; published by the Free Software Foundation; either version 3, or (at
;; your option) any later version.

;; This program is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;; General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;; Boston, MA 02111-1307, USA.

;;; Commentary:

;; Provides support for the :ensure and :pin keywords, which is made available
;; by default by requiring `use-package'.

;;; Code:

(require 'cl-lib)
(require 'use-package-core)

(defgroup use-package-ensure nil
  "Support for :ensure and :pin keywords in use-package declarations."
  :group 'use-package)

(eval-when-compile
  (declare-function package-installed-p "package")
  (declare-function package-read-all-archive-contents "package" ()))

(defcustom use-package-always-ensure nil
  "Treat every package as though it had specified using `:ensure SEXP'.
See also `use-package-defaults', which uses this value."
  :type 'sexp
  :group 'use-package-ensure)

(defcustom use-package-always-pin nil
  "Treat every package as though it had specified using `:pin SYM'.
See also `use-package-defaults', which uses this value."
  :type 'symbol
  :group 'use-package-ensure)

(defcustom use-package-ensure-function #'use-package-ensure-elpa
  "Function that ensures a package is installed.
This function is called with three arguments: the name of the
package declared in the `use-package' form; the arguments passed
to all `:ensure' keywords (always a list, even if only one); and
the current `state' plist created by previous handlers.

Note that this function is called whenever `:ensure' is provided,
even if it is nil. It is up to the function to decide on the
semantics of the various values for `:ensure'.

This function should return non-nil if the package is installed.

The default value uses package.el to install the package."
  :type '(choice (const :tag "package.el" use-package-ensure-elpa)
                 (function :tag "Custom"))
  :group 'use-package-ensure)

;;;; :pin

(defun use-package-normalize/:pin (name keyword args)
  (use-package-only-one (symbol-name keyword) args
    #'(lambda (label arg)
        (cond
         ((stringp arg) arg)
         ((use-package-non-nil-symbolp arg) (symbol-name arg))
         (t
          (use-package-error
           ":pin wants an archive name (a string)"))))))

(eval-when-compile
  (defvar package-pinned-packages)
  (defvar package-archives))

(defun use-package-archive-exists-p (archive)
  "Check if a given ARCHIVE is enabled.

ARCHIVE can be a string or a symbol or 'manual to indicate a
manually updated package."
  (if (member archive '(manual "manual"))
      't
    (let ((valid nil))
      (dolist (pa package-archives)
        (when (member archive (list (car pa) (intern (car pa))))
          (setq valid 't)))
      valid)))

(defun use-package-pin-package (package archive)
  "Pin PACKAGE to ARCHIVE."
  (unless (boundp 'package-pinned-packages)
    (setq package-pinned-packages ()))
  (let ((archive-symbol (if (symbolp archive) archive (intern archive)))
        (archive-name   (if (stringp archive) archive (symbol-name archive))))
    (if (use-package-archive-exists-p archive-symbol)
        (add-to-list 'package-pinned-packages (cons package archive-name))
      (error "Archive '%s' requested for package '%s' is not available."
             archive-name package))
    (unless (bound-and-true-p package--initialized)
      (package-initialize t))))

(defun use-package-handler/:pin (name keyword archive-name rest state)
  (let ((body (use-package-process-keywords name rest state))
        (pin-form (if archive-name
                      `(use-package-pin-package ',(use-package-as-symbol name)
                                                ,archive-name))))
    ;; Pinning should occur just before ensuring
    ;; See `use-package-handler/:ensure'.
    (if (bound-and-true-p byte-compile-current-file)
        (eval pin-form)              ; Eval when byte-compiling,
      (push pin-form body))          ; or else wait until runtime.
    body))

;;;; :ensure

(defvar package-archive-contents)

(defun use-package-ensure-elpa-normalize (args)
  (if (null args)
      (list t)
    (use-package-only-one "ensure" args
      #'(lambda (label arg)
          (cond
           ((symbolp arg)
            (list arg))
           ((and (listp arg) (= 3 (length arg))
                 (symbolp (nth 0 arg))
                 (eq :pin (nth 1 arg))
                 (or (stringp (nth 2 arg))
                     (symbolp (nth 2 arg))))
            (list (cons (nth 0 arg) (nth 2 arg))))
           (t
            (use-package-error
             (concat ":ensure wants an optional package name "
                     "(an unquoted symbol name), or (<symbol> :pin <string>)"))))))))

(defun use-package-ensure-elpa (name args state &optional no-refresh)
  (setq args (use-package-ensure-elpa-normalize args))
  (dolist (ensure args)
    (let ((package
           (or (and (eq ensure t) (use-package-as-symbol name))
               ensure)))
      (when package
        (require 'package)
        (when (consp package)
          (use-package-pin-package (car package) (cdr package))
          (setq package (car package)))
        (unless (package-installed-p package)
          (condition-case-unless-debug err
              (progn
                (when (assoc package (bound-and-true-p
                                      package-pinned-packages))
                  (package-read-all-archive-contents))
                (if (assoc package package-archive-contents)
                    (package-install package)
                  (package-refresh-contents)
                  (when (assoc package (bound-and-true-p
                                        package-pinned-packages))
                    (package-read-all-archive-contents))
                  (package-install package))
                t)
            (error
             (display-warning 'use-package
                              (format "Failed to install %s: %s"
                                      name (error-message-string err))
                              :error))))))))

;;;###autoload
(defun use-package-normalize/:ensure (name keyword args)
  args)

;;;###autoload
(defun use-package-handler/:ensure (name keyword ensure rest state)
  (let* ((body (use-package-process-keywords name rest state)))
    ;; We want to avoid installing packages when the `use-package' macro is
    ;; being macro-expanded by elisp completion (see `lisp--local-variables'),
    ;; but still install packages when byte-compiling, to avoid requiring
    ;; `package' at runtime.
    (if (bound-and-true-p byte-compile-current-file)
        ;; Eval when byte-compiling,
        (funcall use-package-ensure-function name ensure state)
      ;;  or else wait until runtime.
      (push `(,use-package-ensure-function ',name ',ensure ',state)
            body))
    body))

(add-to-list 'use-package-defaults
             '(:ensure (list use-package-always-ensure)
                       (lambda (name args)
                         (and use-package-always-ensure
                              (not (plist-member args :load-path))))) t)

(add-to-list 'use-package-defaults
             '(:pin use-package-always-pin use-package-always-pin) t)

(add-to-list 'use-package-keywords :ensure)
(add-to-list 'use-package-keywords :pin)

(provide 'use-package-ensure)

;;; use-package-ensure.el ends here
